frame/balances/src/lib.rs

```diff
@@ -492,1058 +628,998 @@ decl_module! {
-	fn extend_lock(
-		id: LockIdentifier,
-		who: &T::AccountId,
-		amount: T::Balance,
-		reasons: WithdrawReasons,
-	) -> DispatchResult {
-		if amount.is_zero() || reasons.is_empty() {
-			return Ok(());
+	// A value placed in storage that represents the current version of the Balances storage.
+	// This value is used by the `on_runtime_upgrade` logic to determine whether we run
+	// storage migration logic. This should match directly with the semantic versions of the Rust crate.
+	#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug)]
+	pub enum Releases {
+		V1_0_0,
+		V2_0_0,
+	}
+	impl Default for Releases {
+		fn default() -> Self {
+			Releases::V1_0_0
 		}
-
-		let mut new_lock = Some(BalanceLock {
-			id,
-			lock_for: LockFor::Common { amount },
-			lock_reasons: reasons.into(),
-		});
-		let mut poisoned = false;
-		let mut locks = Self::locks(who)
-			.into_iter()
-			.filter_map(|l| {
-				if l.id == id {
-					if let LockFor::Common { amount: a } = l.lock_for {
-						new_lock.take().map(|nl| BalanceLock {
-							id: l.id,
-							lock_for: {
-								match nl.lock_for {
-									// Only extend common lock type
-									LockFor::Common { amount: na } => LockFor::Common {
-										amount: (a).max(na),
-									},
-									// Not allow to extend other combination/type lock
-									//
-									// And the lock is always with lock id
-									// it's impossiable to match a (other lock, common lock)
-									// under this if condition
-									_ => {
-										poisoned = true;
-
-										nl.lock_for
-									}
-								}
-							},
-							lock_reasons: l.lock_reasons | nl.lock_reasons,
-						})
-					} else {
-						Some(l)
+	}
+}
+pub use pallet::{imbalances::*, *};
+
+pub mod migration {
+	#[cfg(feature = "try-runtime")]
+	pub mod try_runtime {
+		// --- substrate ---
+		use frame_support::{pallet_prelude::*, traits::StorageInstance};
+		// --- darwinia ---
+		use crate::*;
+
+		macro_rules! generate_storage_types {
+			($prefix:expr, $storage:expr, $name:ident => Value<$value:ty>) => {
+				paste::paste! {
+					type $name = StorageValue<[<$name Instance>], $value, ValueQuery>;
+
+					struct [<$name Instance>];
+					impl StorageInstance for [<$name Instance>] {
+						const STORAGE_PREFIX: &'static str = $storage;
+
+						fn pallet_prefix() -> &'static str { $prefix }
 					}
-				} else {
-					Some(l)
 				}
-			})
-			.collect::<Vec<_>>();
-
-		if poisoned {
-			Err(<Error<T, I>>::LockP)?;
-		}
-
-		if let Some(lock) = new_lock {
-			locks.push(lock)
+			};
 		}
 
-		Self::update_locks(who, &locks);
+		generate_storage_types!("Instance0DarwiniaBalances", "TotalIssuance", OldRingTotalIssuance => Value<()>);
+		generate_storage_types!("Balances", "TotalIssuance", NewRingTotalIssuance => Value<()>);
+		generate_storage_types!("Instance1DarwiniaBalances", "TotalIssuance", OldKtonTotalIssuance => Value<()>);
+		generate_storage_types!("Kton", "TotalIssuance", NewKtonTotalIssuance => Value<()>);
```

Comment:
No need to do the migration for `Account`, `Locks`?

```rs
/// The balance of an account.
///
/// NOTE: This is only used in the case that this pallet is used to store balances.
#[pallet::storage]
pub type Account<T: Config<I>, I: 'static = ()> =
	StorageMap<_, Blake2_128Concat, T::AccountId, T::BalanceInfo, ValueQuery>;

/// Any liquidity locks on some account balances.
/// NOTE: Should only be accessed when setting, changing and freeing a lock.
#[pallet::storage]
#[pallet::getter(fn locks)]
pub type Locks<T: Config<I>, I: 'static = ()> = StorageMap<
	_,
	Blake2_128Concat,
	T::AccountId,
	Vec<BalanceLock<T::Balance, T::BlockNumber>>,
	ValueQuery,
>;
```

Reply:
Just checking.

And I skip the map items checking.